workshops/web_chat/README.md

```diff
@@ -0,0 +1,782 @@
+
+```js
+import { serve } from "https://deno.land/std@0.75.0/http/server.ts"
+
+const server = serve(":8080")
+console.log(`Chat server is running on 8080`)
+
+for await (const req of server) {
+    if (req.url === "/") {
+        req.url = "/index.html"
+    }
+    try {
+        const data = await Deno.readTextFile(req.url.substring(1))
+        await req.respond({ status: 200, body: data })
+    } catch {
+        await req.respond({ status: 404 })
+    }
+}
+```
+
+Here's what's happening here:
+
+- By default, all requests go to "/"
+- If the request is going to "/", we want to set the request URL to "/index.html" -- we want to serve index.html by default
+- `req.url.substring(1)` removes the first character from `req.url`. Therefore if `req.url` is "/index.html", `req.url.substring(1)` will be "index.html"
+- We read the file requested by the user, and return that file
+- If that file does not exist, we send a 404 error.
+
+> Repl.it TIP:
+> ![Click format and reload on top center of page to format your code and reload your page preview](https://cloud-pq5lbfiab.vercel.app/4formatreload.png)
+
+
+You should see the UI created earlier. Let's create a websocket that a webpage can connect to:
+
+```js
+import { serve } from "https://deno.land/std@0.75.0/http/server.ts"
+import { acceptWebSocket, WebSocket } from "https://deno.land/std@0.75.0/ws/mod.ts"
+
+const server = serve(":8080")
+console.log(`Chat server is running on 8080`)
+
+for await (const req of server) {
+    try {
+        const { conn, r: bufReader, w: bufWriter, headers } = req
+        let socket = await acceptWebSocket({
+            conn,
+            bufReader,
+            bufWriter,
+            headers,
+        })
+    } catch (error) {
+        if (req.url === "/") {
+            req.url = "/index.html"
+        }
+        try {
+            const data = await Deno.readTextFile(req.url.substring(1))
+            await req.respond({ status: 200, body: data })
+        } catch {
+            await req.respond({ status: 404 })
+        }  
+    }
+}
+```
+
+Every time someone vists our page, we'll want to send index.html to them. However, trying to connect to a websocket also counts as a request. For every request that we get, we want to try to parse it as a websocket request. If that fails, then we send the index.html. If you hit "Run" now, you should see... no changes. This is because we haven't added any code to connect to our websocket in the frontend yet! Add the following code to frontend.js:
+
+```js
+let ws
+
+document.addEventListener("DOMContentLoaded", _ => {
+    ws = new WebSocket(`wss://${window.location.host}`)
+    ws.onopen = function (event) {
+        console.log("Socket connection is open")
+        ws.send(JSON.stringify({
+            type: "open",
+            data: {}
+        }))
+    }
+
+    ws.onmessage = function (event) {
+      console.log("Message received")
```

Comment:
So far you've been indenting with 4 spaces or with a tab, but it looks like the indentation is 2 spaces here. Can you go through and make sure all indentation is consistent?